### PR TITLE
Improve codegen stability

### PR DESCRIPTION
**Pull request by karthijrk**
In this PR, we addressed the following issues that came out as a result of improving the codegen stability.

1. Introduced guc to enable / disable generators. 
```
codegen_exec_variable_list;
codegen_slot_getattr;
codegen_exec_eval_expression;
codegen_advance_aggregate;
```
2. `slot_getattr` is moved to .h with 8.3 merge. This cause issue when we register as external function. LLVM expect the name of the external function to be unique for each pointer. In order to fix it, we introduced wrapper function in codegen_wrapper.cc and use that instead of slot_getattr.

3. In `slot_getattr_codegen`, we found that IR code for logic `!(Expr1 & Expr2)` and `hasnull` is wrong. We fixed in this PR. 

4. We fixed the issue of `slot_getattr` function pointer being null for expression evaluation code generation.

5. Cpplint was red due to a couple of violation in `advance_aggregate`. This has been fixed in this PR.

Existing ICG `join.sql` covers pretty much what we fixed here.

@foyzur @hardikar @armenatzoglou @xinzweb Please take a look when you get a chance.

### CODE REVIEW

File: src/backend/codegen/exec_eval_expr_codegen.cc [hunk 1]
```diff
@@ -122,7 +122,7 @@ bool ExecEvalExprCodegen::GenerateExecEvalExpr(
   // we revert to use the external slot_getattr()
   if (nullptr == slot_getattr_codegen_) {
     gen_info_.llvm_slot_getattr_func =
-        codegen_utils->GetOrRegisterExternalFunction(slot_getattr,
+        codegen_utils->GetOrRegisterExternalFunction(slot_getattr_regular,
```
Comment (danielgustafsson) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: I'm curious, what was the error/warning this induced? `slot_getattr` has been inlined in the headerfile since well before the 8.3 merge (and upstream does Not have it in the headerfile, it's in `src/backend/access/common/heaptuple.c`)

File: src/backend/codegen/codegen_wrapper.cc [hunk 2]
```diff
@@ -104,51 +102,25 @@ void SetActiveCodeGeneratorManager(void* manager) {
...
 
 void* ExecVariableListCodegenEnroll(
     ExecVariableListFn regular_func_ptr,
     ExecVariableListFn* ptr_to_chosen_func_ptr,
     ProjectionInfo* proj_info,
     TupleTableSlot* slot) {
-  ExecVariableListCodegen* generator = CodegenEnroll<ExecVariableListCodegen>(
-      regular_func_ptr, ptr_to_chosen_func_ptr, proj_info, slot);
+  CodegenManager* manager = static_cast<CodegenManager*>(
+      GetActiveCodeGeneratorManager());
```
Comment (armenatzoglou) [inline src/backend/codegen/codegen_wrapper.cc]: Previously we were assuming that the enroller would always use active manager. What changed? Can we not pass the active manager and still use the active manager as before from inside the enroller?

File: src/backend/codegen/include/codegen/codegen_config.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
+//---------------------------------------------------------------------------
+//  Greenplum Database
+//  Copyright (C) 2016 Pivotal Software, Inc.
+//
+//  @filename:
+//    codegen_interface.h
```
Comment (armenatzoglou) [inline src/backend/codegen/include/codegen/codegen_config.h]: Wrong filename

File: src/backend/utils/misc/guc_gp.c [hunk 4]
```diff
@@ -3352,6 +3356,62 @@ struct config_bool ConfigureNamesBool_gp[] =
...
+		&codegen_exec_variable_list,
+#ifdef USE_CODEGEN
+		true,
+#else
+		false,
+#endif
+		assign_codegen, NULL
+	},
+	{
+			{"codegen_slot_getattr", PGC_USERSET, DEVELOPER_OPTIONS,
```
Comment (armenatzoglou) [inline src/backend/utils/misc/guc_gp.c]: Indentation

File: src/backend/codegen/exec_eval_expr_codegen.cc (hunk 1 above)
Comment (hardikar) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: If I remember correctly, it was in a C file. We added #ifdef guards for the `static` component, such that when GPDB is built with codegen, it becomes non-static - and so there only one implementation (address) in the final binary. Once it's moved to the header file, each translation unit gets its own "implementation" of that function and thus address. 

Our logic assumes each function is associated with a unique function address. But moving it to a header file means that is no longer the case. This is a tricky point - and suggests we should reconsider this mechanism - and we have a story for that already.

Also, I don't think this gave any warnings or errors in the build. ICG with codegen just was red - and we got a runtime assertion failure from LLVM that a function with the same name was already added as a global mapping.

File: src/backend/codegen/codegen_wrapper.cc (hunk 2 above)
Comment (karthijrk) [inline src/backend/codegen/codegen_wrapper.cc]: I don't want to create dependency that codegen_manager.h depend on codegen_wrapper.h. Hence I made the codegen_manager api to take manager as an argument.

File: src/backend/codegen/include/codegen/codegen_config.h (hunk 3 above)
Comment (xinzweb) [inline src/backend/codegen/include/codegen/codegen_config.h]: Fixed. Thanks.

File: src/backend/codegen/include/codegen/codegen_config.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+//    Configure Class to handle gucs in gpdb
+//
+//---------------------------------------------------------------------------
+#ifndef GPCODEGEN_CODEGEN_CONFIG_H_  // NOLINT(build/header_guard)
+#define GPCODEGEN_CODEGEN_CONFIG_H_
+
+extern "C" {
+// Variables that are defined in guc
+extern bool codegen;
+extern bool init_codegen;
```
Comment (xinzweb) [inline src/backend/codegen/include/codegen/codegen_config.h]: No need to dup the comments, it's already in `guc_gp.c`

File: src/backend/codegen/exec_eval_expr_codegen.cc [hunk 6]
```diff
@@ -120,14 +120,15 @@ bool ExecEvalExprCodegen::GenerateExecEvalExpr(
 
   // In case the generation above either failed or was not needed,
```
Comment (hardikar) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: Reading these comments - I realized they don't quite make sense - can you update these with you changes.

File: src/backend/codegen/CMakeLists.txt [hunk 7]
```diff
@@ -264,7 +264,8 @@ function(add_cmockery_gtest TEST_NAME TEST_SOURCES)
     )
     target_include_directories(${TEST_NAME} PUBLIC ${TEST_LIB_INC_DIRECTORIES})
     # Bring these from $ENV{LIBS}
-    target_link_libraries(${TEST_NAME} "-ldl -lbz2 -lxml2 -lz -lm -lcurl -L../../port -lpgport_srv" gpcodegen gtest)
+    # -ldl -lbz2 -lxml2 -lssl -lcrypto -lz -lm -lcurl -lldap
```
Comment (hardikar) [inline src/backend/codegen/CMakeLists.txt]: remove this comment.

File: src/backend/codegen/CMakeLists.txt [hunk 8]
```diff
@@ -264,7 +264,8 @@ function(add_cmockery_gtest TEST_NAME TEST_SOURCES)
...
+    target_link_libraries(${TEST_NAME} "-lssl -lcrypto -lldap -ldl -lbz2 -lxml2 -lz -lm -lcurl -L../../port -lpgport_srv" gpcodegen gtest)
```
Comment (hardikar) [inline src/backend/codegen/CMakeLists.txt]: This list of libraries should be derived from configure, instead of being hard coded. That way it should work no matter the configuration.

File: src/backend/codegen/include/codegen/codegen_manager.h [hunk 9]
```diff
@@ -47,6 +49,49 @@ class CodegenManager {
...
+   * @return Pointer to ClassType
+   **/
+  template <typename ClassType, typename FuncType, typename ...Args>
+  static ClassType* CreateAndEnrollGenerator(
+      CodegenManager* manager,
+      FuncType regular_func_ptr,
+      FuncType* ptr_to_chosen_func_ptr,
+      Args&&... args) {  // NOLINT(build/c++11)
+    bool can_enroll =
+        (nullptr != manager) &&
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/codegen_manager.h]: Can you describe here (as comments) the scenarios that can cause this function call with manager = nullptr.

File: src/backend/utils/misc/guc_gp.c [hunk 10]
```diff
@@ -3352,6 +3356,62 @@ struct config_bool ConfigureNamesBool_gp[] =
...
 		assign_codegen, NULL
 	},
+	{
+		{"codegen_exec_variable_list", PGC_USERSET, DEVELOPER_OPTIONS,
+			gettext_noop("Enable codegen for ExecVariableList"),
+			NULL,
+			GUC_NO_SHOW_ALL | GUC_NOT_IN_SAMPLE | GUC_GPDB_ADDOPT
+		},
+		&codegen_exec_variable_list,
+#ifdef USE_CODEGEN
```
Comment (hardikar) [inline src/backend/utils/misc/guc_gp.c]: Well it's an integer GUC unused in GPDB, and honestly the value of 0 doesn't really disable any functionality - that's handled by some other GUC, so it really doesn't matter if USE_CODEGEN guard is present or not.

File: src/backend/codegen/exec_eval_expr_codegen.cc (hunk 1 above)
Comment (danielgustafsson) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: `slot_getattr()` was moved to the headerfile at some point before the open sourcing of GPDB and not as part of the 8.3 merge. I assume this was for a performance benefit (as it's inlined) but it's not documented. Upstream has it in the .c file, have you tried moving this back to avoid the complexity? I'm all for minimizing the diff to upstream and I'm not convinced that marking this inline carries a huge performance boost. Well worth pursuing IMHO.

File: src/backend/codegen/slot_getattr_codegen.cc [hunk 11]
```diff
@@ -440,16 +451,21 @@ bool SlotGetAttrCodegen::GenerateSlotGetAttr(
...
+
+      // !(Expr_1 & Expr_2)
+      llvm::Value* llvm_att_isnull = irb->CreateNot(llvm_expr1_and_expr2,
+                                                    "llvm_att_isnull");
       // }}
 
-      llvm::Value* llvm_hasnulls = irb->CreateTrunc(
+      llvm::Value* llvm_hasnulls = irb->CreateICmpNE(
           irb->CreateAnd(llvm_heaptuple_t_data_t_infomask,
                          codegen_utils->GetConstant<uint16>(HEAP_HASNULL)),
```
Comment (danielgustafsson) [inline src/backend/codegen/slot_getattr_codegen.cc]: Note that `src/include/c.h` typedefs int16, int64 etc. Since gpdb doesn't use the standard _t types it can be quite dangerous to assume they are equal (for example, int64_t isn't compatible with int64 in gpdb). Using the same definitions is crucial and can be achieved by pulling in the same autoconf macros as `c.h` if including that isn't an option in the c++ code.

File: src/backend/codegen/CMakeLists.txt (hunk 8 above)
Comment (danielgustafsson) [inline src/backend/codegen/CMakeLists.txt]: +1, autoconf must drive the configuration of the entire tree.

File: src/backend/codegen/exec_eval_expr_codegen.cc (hunk 1 above)
Comment (karthijrk) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: oh. That's great. I agree with you. We want to keep the diff with the upstream minimum. In that case, I will move the definition of `slot_getattr` back to `heaptuple.c` similar to Postgres.

File: src/backend/utils/misc/guc_gp.c (hunk 4 above)
Comment (karthijrk) [inline src/backend/utils/misc/guc_gp.c]: Fixed. Thanks good catch.

File: src/backend/codegen/exec_eval_expr_codegen.cc [hunk 12]
```diff
@@ -122,7 +122,7 @@ bool ExecEvalExprCodegen::GenerateExecEvalExpr(
   // we revert to use the external slot_getattr()
   if (nullptr == slot_getattr_codegen_) {
     gen_info_.llvm_slot_getattr_func =
-        codegen_utils->GetOrRegisterExternalFunction(slot_getattr,
+        codegen_utils->GetOrRegisterExternalFunction(slot_getattr_regular,
                                                      "slot_getattr");
```
Comment (karthijrk) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: I am going to revert the change from `slot_getattr` to `slot_getattr_regular`.

File: src/backend/codegen/slot_getattr_codegen.cc [hunk 13]
```diff
@@ -65,6 +67,13 @@ SlotGetAttrCodegen* SlotGetAttrCodegen::GetCodegenInstance(
     TupleTableSlot *slot,
     int max_attr) {
 
+  // TODO(krajaraman, frahman) : Refactor so creation happens through
```
Comment (karthijrk) [inline src/backend/codegen/slot_getattr_codegen.cc]: Done :)

File: src/backend/codegen/slot_getattr_codegen.cc (hunk 11 above)
Comment (karthijrk) [inline src/backend/codegen/slot_getattr_codegen.cc]: Let's go with int64. C++ introduced `int64_t` later and they decided to keep the suffix `_t` to avoid name collision. I agree with Daniel lets keep it as uint64

File: src/backend/utils/misc/guc_gp.c [hunk 14]
```diff
@@ -577,6 +577,10 @@ bool		optimizer_array_constraints;
 bool		init_codegen;
 bool		codegen;
 bool		codegen_validate_functions;
+bool		codegen_exec_variable_list;
+bool		codegen_slot_getattr;
+bool		codegen_exec_eval_expression;
```
Comment (karthijrk) [inline src/backend/utils/misc/guc_gp.c]: Done :)

File: src/backend/codegen/CMakeLists.txt (hunk 8 above)
Comment (karthijrk) [inline src/backend/codegen/CMakeLists.txt]: I tried to see if I can fix this quickly but seems like it needs more work. Found that configure doesn't set the linker flag in which case it would have been easier to plumb it to codegen's CMake. I created a separate story in our QX backlog to address this debt. 
Here is the story link -- #131758529

**Comment by karthijrk** [comment]
I think I have addressed pretty much all comments. Let me know if there is any other concern.

File: src/backend/codegen/slot_getattr_codegen.cc (hunk 11 above)
Comment (hardikar) [inline src/backend/codegen/slot_getattr_codegen.cc]: +1 for consistency between C and C++ type usage in GPDB

File: src/backend/codegen/include/codegen/codegen_manager.h [hunk 15]
```diff
@@ -47,6 +49,49 @@ class CodegenManager {
   ~CodegenManager() = default;
 
   /**
+   * @brief Template function to facilitate enroll for any type of
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/codegen_manager.h]: Let's work on this tomorrow - I'd really like a good comment here (or on the manager) because it's the entry way to the GPDB codegen framework.

File: src/backend/codegen/slot_getattr_codegen.cc [hunk 16]
```diff
@@ -440,16 +445,21 @@ bool SlotGetAttrCodegen::GenerateSlotGetAttr(
...
+      llvm::Value* llvm_att_isnull = irb->CreateNot(llvm_expr1_and_expr2,
+                                                    "llvm_att_isnull");
       // }}
 
-      llvm::Value* llvm_hasnulls = irb->CreateTrunc(
+      llvm::Value* llvm_hasnulls = irb->CreateICmpNE(
           irb->CreateAnd(llvm_heaptuple_t_data_t_infomask,
                          codegen_utils->GetConstant<uint16>(HEAP_HASNULL)),
-                         codegen_utils->GetType<bool>());
+                         codegen_utils->GetConstant<uint16_t>(0),
```
Comment (hardikar) [inline src/backend/codegen/slot_getattr_codegen.cc]: s/uint16_t/uint16/

File: src/backend/codegen/exec_eval_expr_codegen.cc (hunk 1 above)
Comment (danielgustafsson) [inline src/backend/codegen/exec_eval_expr_codegen.cc]: :+1:
